File: ports/ts/test/rules.test.js
"use strict";
const test = require("node:test");
const assert = require("node:assert");
const { analyzeManifest, run, SENSITIVE_PERMISSIONS } = require("../dist/rules.js");

function byTitle(fs) {
  const m = {};
  for (const f of fs) m[f.title] = f;
  return m;
}

test("debuggable is HIGH", () => {
  const fs = analyzeManifest({ debuggable: true });
  assert.equal(byTitle(fs)["Application is debuggable"].severity, "HIGH");
});

test("allow_backup is MEDIUM", () => {
  const fs = analyzeManifest({ allow_backup: true });
  assert.ok(byTitle(fs)["ADB backup allowed"]);
});

test("cleartext true flagged", () => {
  const fs = analyzeManifest({ uses_cleartext_traffic: true });
  assert.ok(byTitle(fs)["Cleartext (HTTP) traffic permitted"]);
});

test("cleartext false suppresses NSC", () => {
  const fs = analyzeManifest({ target_sdk: 33, uses_cleartext_traffic: false });
  assert.ok(!byTitle(fs)["No Network Security Config"]);
});

test("NSC flagged when target high and unset", () => {
  const fs = analyzeManifest({ target_sdk: 30 });
  assert.ok(byTitle(fs)["No Network Security Config"]);
});

test("NSC present suppresses", () => {
  const fs = analyzeManifest({ target_sdk: 30, network_security_config: "@xml/nsc" });
  assert.ok(!byTitle(fs)["No Network Security Config"]);
});

test("exported with filter is MEDIUM", () => {
  const fs = analyzeManifest({
    components: [{ kind: "activity", name: ".Main", exported: true, has_permission: false, intent_filters: 1 }],
  });
  assert.equal(byTitle(fs)["Exported activity without permission: .Main"].severity, "MEDIUM");
});

test("exported no filter is LOW", () => {
  const fs = analyzeManifest({
    components: [{ kind: "service", name: ".Sync", exported: true, has_permission: false, intent_filters: 0 }],
  });
  assert.equal(byTitle(fs)["Exported service without permission: .Sync"].severity, "LOW");
});

test("guarded component not flagged", () => {
  const fs = analyzeManifest({
    components: [{ kind: "activity", name: ".G", exported: true, has_permission: true, intent_filters: 1 }],
  });
  assert.equal(fs.length, 0);
});

test("sensitive permissions sorted unique", () => {
  const fs = analyzeManifest({
    permissions: [
      "android.permission.READ_SMS",
      "android.permission.CAMERA",
      "android.permission.READ_SMS",
      "android.permission.INTERNET",
    ],
  });
  const ev = fs.filter((f) => f.title.startsWith("Sensitive permission")).map((f) => f.evidence);
  assert.deepEqual(ev, ["android.permission.CAMERA", "android.permission.READ_SMS"]);
});

test("low minSdk flagged", () => {
  const fs = analyzeManifest({ min_sdk: 19 });
  assert.ok(byTitle(fs)["Low minSdkVersion (19)"]);
});

test("high minSdk not flagged", () => {
  const fs = analyzeManifest({ min_sdk: 28 });
  assert.ok(!byTitle(fs)["Low minSdkVersion (28)"]);
});

test("hardened manifest is clean", () => {
  const fs = analyzeManifest({
    package: "com.acme.app",
    min_sdk: 28,
    target_sdk: 33,
    allow_backup: false,
    network_security_config: "@xml/nsc",
    uses_cleartext_traffic: false,
  });
  assert.equal(fs.length, 0);
});

test("run emits JSON array", () => {
  const out = run(JSON.stringify({ package: "com.x", debuggable: true, allow_backup: false }));
  const fs = JSON.parse(out);
  assert.equal(fs.length, 1);
  assert.equal(fs[0].severity, "HIGH");
});

test("run empty is empty array", () => {
  const out = run(JSON.stringify({ package: "com.x", allow_backup: false }));
  assert.deepEqual(JSON.parse(out), []);
});

test("sensitive set has 10 entries", () => {
  assert.equal(SENSITIVE_PERMISSIONS.size, 10);
});

File: ports/go/rules_test.go
package main

import (
	"bytes"
	"encoding/json"
	"strings"
	"testing"
)

func boolPtr(b bool) *bool { return &b }

func titles(fs []Finding) map[string]Finding {
	m := map[string]Finding{}
	for _, f := range fs {
		m[f.Title] = f
	}
	return m
}

func TestDebuggable(t *testing.T) {
	fs := AnalyzeManifest(Manifest{Debuggable: true})
	f, ok := titles(fs)["Application is debuggable"]
	if !ok || f.Severity != "HIGH" {
		t.Fatalf("expected HIGH debuggable finding, got %#v", fs)
	}
}

func TestAllowBackup(t *testing.T) {
	fs := AnalyzeManifest(Manifest{AllowBackup: true})
	if _, ok := titles(fs)["ADB backup allowed"]; !ok {
		t.Fatal("expected ADB backup finding")
	}
}

func TestCleartextTrue(t *testing.T) {
	fs := AnalyzeManifest(Manifest{UsesCleartextTraffic: boolPtr(true)})
	if _, ok := titles(fs)["Cleartext (HTTP) traffic permitted"]; !ok {
		t.Fatal("expected cleartext finding")
	}
}

func TestCleartextFalseSuppressesNSC(t *testing.T) {
	fs := AnalyzeManifest(Manifest{TargetSDK: 33, UsesCleartextTraffic: boolPtr(false)})
	if _, ok := titles(fs)["No Network Security Config"]; ok {
		t.Fatal("cleartext=false should suppress NSC finding")
	}
}

func TestNSCWhenTargetHighAndUnset(t *testing.T) {
	fs := AnalyzeManifest(Manifest{TargetSDK: 30})
	if _, ok := titles(fs)["No Network Security Config"]; !ok {
		t.Fatal("expected NSC finding")
	}
}

func TestNSCNotFlaggedWhenConfigured(t *testing.T) {
	fs := AnalyzeManifest(Manifest{TargetSDK: 30, NetworkSecurityConfig: "@xml/nsc"})
	if _, ok := titles(fs)["No Network Security Config"]; ok {
		t.Fatal("NSC present should suppress finding")
	}
}

func TestExportedComponentWithFilter(t *testing.T) {
	fs := AnalyzeManifest(Manifest{Components: []Component{
		{Kind: "activity", Name: ".Main", Exported: true, IntentFilters: 1},
	}})
	f, ok := titles(fs)["Exported activity without permission: .Main"]
	if !ok || f.Severity != "MEDIUM" {
		t.Fatalf("expected MEDIUM exported finding, got %#v", fs)
	}
}

func TestExportedComponentNoFilter(t *testing.T) {
	fs := AnalyzeManifest(Manifest{Components: []Component{
		{Kind: "service", Name: ".Sync", Exported: true, IntentFilters: 0},
	}})
	f := titles(fs)["Exported service without permission: .Sync"]
	if f.Severity != "LOW" {
		t.Fatalf("expected LOW, got %q", f.Severity)
	}
}

func TestGuardedComponentNotFlagged(t *testing.T) {
	fs := AnalyzeManifest(Manifest{Components: []Component{
		{Kind: "activity", Name: ".G", Exported: true, HasPermission: true, IntentFilters: 1},
	}})
	if len(fs) != 0 {
		t.Fatalf("guarded exported component should not be flagged, got %#v", fs)
	}
}

func TestSensitivePermissionSortedUnique(t *testing.T) {
	fs := AnalyzeManifest(Manifest{Permissions: []string{
		"android.permission.READ_SMS", "android.permission.CAMERA",
		"android.permission.READ_SMS", "android.permission.INTERNET",
	}})
	var got []string
	for _, f := range fs {
		if strings.HasPrefix(f.Title, "Sensitive permission") {
			got = append(got, f.Evidence)
		}
	}
	want := []string{"android.permission.CAMERA", "android.permission.READ_SMS"}
	if len(got) != len(want) || got[0] != want[0] || got[1] != want[1] {
		t.Fatalf("want sorted unique %v, got %v", want, got)
	}
}

func TestLowMinSdk(t *testing.T) {
	fs := AnalyzeManifest(Manifest{MinSDK: 19})
	if _, ok := titles(fs)["Low minSdkVersion (19)"]; !ok {
		t.Fatal("expected low minSdk finding")
	}
}

func TestHighMinSdkNotFlagged(t *testing.T) {
	fs := AnalyzeManifest(Manifest{MinSDK: 28})
	if _, ok := titles(fs)["Low minSdkVersion (28)"]; ok {
		t.Fatal("minSdk 28 should not be flagged")
	}
}

func TestHardenedManifestClean(t *testing.T) {
	fs := AnalyzeManifest(Manifest{
		Package: "com.acme.app", MinSDK: 28, TargetSDK: 33,
		AllowBackup: false, NetworkSecurityConfig: "@xml/nsc",
		UsesCleartextTraffic: boolPtr(false),
	})
	if len(fs) != 0 {
		t.Fatalf("hardened manifest should be clean, got %#v", fs)
	}
}

func TestRunJSON(t *testing.T) {
	in := strings.NewReader(`{"package":"com.x","debuggable":true,"allow_backup":false}`)
	var out bytes.Buffer
	if err := Run(in, &out); err != nil {
		t.Fatal(err)
	}
	var fs []Finding
	if err := json.Unmarshal(out.Bytes(), &fs); err != nil {
		t.Fatal(err)
	}
	if len(fs) != 1 || fs[0].Severity != "HIGH" {
		t.Fatalf("expected one HIGH finding, got %#v", fs)
	}
}

func TestRunEmptyArrayNotNull(t *testing.T) {
	in := strings.NewReader(`{"package":"com.x","allow_backup":false}`)
	var out bytes.Buffer
	if err := Run(in, &out); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(out.String(), "[]") {
		t.Fatalf("expected empty JSON array, got %s", out.String())
	}
}

func TestItoa(t *testing.T) {
	cases := map[int]string{0: "0", 7: "7", 24: "24", -5: "-5", 1000: "1000"}
	for n, want := range cases {
		if got := itoa(n); got != want {
			t.Fatalf("itoa(%d)=%q want %q", n, got, want)
		}
	}
}
